### PR TITLE
Block producer swap bot

### PR DESCRIPTION
**Pull request by Art17**
Bot for processing swap requests on Ethereum and sending swapped tokens on remprotocol.

### CODE REVIEW

File: block_producer_swap_bot/README.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+```bash
+$ nano cli/config.ini
+```
+
+Paste into the config file following content:
+
+```
+[NODES]
+cleos=cleos --url http://127.0.0.1:8888
```
Comment (olexiybuyanskyy) [inline block_producer_swap_bot/README.md]: should we use remcli here?

File: block_producer_swap_bot/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+eth-provider=wss://rinkeby.infura.io/ws/v3/<your infura id>
+[REMIO]
+block-producer-permission=<your account name>@active
+```
+
+Replace cleos, eth-provider, block-producer-permission with your cleos path and options, link to Ethereum node with websocket connection and your block producer's account for signing .
```
Comment (olexiybuyanskyy) [inline block_producer_swap_bot/README.md]: cleos -> remcli

File: block_producer_swap_bot/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+```bash
+$ python3 block_producer_swap_bot.py eth-swap-bot process-swaps
+```
+
+## Development
+
+Manually process particular swap transaction
+```bash
+$ cleos --url http://127.0.0.1:8000 --wallet-url http://127.0.0.1:6666 push action remio.swap processswap '[ "producer111a", "0x30a9479fc792d3219aba23440235a4a7e4ab32e7ff86a08d878778c5076d206b", "1c6ae7719a2a3b4ecb19584a30ff510ba1b6ded86e1fd8b8fc22f1179c622a32", "EOS7oNmmxo8yh8gmYLUGNCwNAFfLmrMxtmrzmFPG29CpGm5Bq4FGC", "20.0000 REM", "2019-07-25T17:51:47" ]' -p producer111a@active
```
Comment (olexiybuyanskyy) [inline block_producer_swap_bot/README.md]: cleos -> remcli
remio.swap -> rem.swap

File: block_producer_swap_bot/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+```
+
+Manually finish swap
+```bash
+cleos --url http://127.0.0.1:8000 --wallet-url http://127.0.0.1:6666 push action remio.swap finishswap '[ "receiver", "0x30a9479fc792d3219aba23440235a4a7e4ab32e7ff86a08d878778c5076d206b", "1c6ae7719a2a3b4ecb19584a30ff510ba1b6ded86e1fd8b8fc22f1179c622a32", "EOS7oNmmxo8yh8gmYLUGNCwNAFfLmrMxtmrzmFPG29CpGm5Bq4FGC","20.0000 REM", "2019-07-25T17:51:47", "SIG_K1_AnEQt8cs8Uscfg3FhRM3za3WxzzP8VHcUCAhfwoQYUb8EFbEujrTswAeiCaNWKzSfADWzVBoBv3mxV7qb4ymo4QHXFfTCLQWd6kH45WL2KggUP4aTpvH", "EOS7oNmmxo8yh8gmYLUGNCwNAFfLmrMxtmrzmFPG29CpGm5Bq4FGC", "EOS7oNmmxo8yh8gmYLUGNCwNAFfLmrMxtmrzmFPG29CpGm5Bq4FGC" ]' -p producer111a@active
```
Comment (olexiybuyanskyy) [inline block_producer_swap_bot/README.md]: cleos -> remcli
remio.swap -> rem.swap

**Review by olexiybuyanskyy — COMMENTED** [review]
I see many other places where cleos used instead of remcli

File: block_producer_swap_bot/cli/eth_swap_bot/service.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+        return result
+
+    @staticmethod
+    def get_event_block_number(event):
+        return event.get('blockNumber')
+
+    def wait_for_event_confirmation(self, event):
+        while True:
+            event_block_number = EthSwapBot.get_event_block_number(event)
```
Comment (roman-tik) [inline block_producer_swap_bot/cli/eth_swap_bot/service.py:93]: we also need to handle the case when eth chain gets reorganized and the transaction gets orphaned and replaced while we count the confirmations

File: block_producer_swap_bot/cli/eth_swap_bot/service.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                sleep(SHORT_POLLING_CONFIRMATION_INTERVAL)
+            else:
+                break
+
+    def handle_swap_request_event(self, event):
+        result = self.get_swap_request_parameters(event)
+
+        txid = result.get('txid')
+        timestamp = result.get('timestamp')
+        result['timestamp'] = datetime.utcfromtimestamp(timestamp).strftime("%Y-%m-%dT%H:%M:%S")
```
Comment (roman-tik) [inline block_producer_swap_bot/cli/eth_swap_bot/service.py:105]: as agreed, the timestamp has to be a `decimal` number in the epoch seconds

**Review by roman-tik — CHANGES_REQUESTED** [review]

File: block_producer_swap_bot/README.md (hunk 1 above)
Comment (Art17) [reply block_producer_swap_bot/README.md]: Done.

File: block_producer_swap_bot/README.md (hunk 2 above)
Comment (Art17) [reply block_producer_swap_bot/README.md]: Done.

File: block_producer_swap_bot/README.md (hunk 3 above)
Comment (Art17) [reply block_producer_swap_bot/README.md]: Done.

File: block_producer_swap_bot/cli/eth_swap_bot/service.py (hunk 6 above)
Comment (Art17) [reply block_producer_swap_bot/cli/eth_swap_bot/service.py:105]: Nodeos converts it under the hood.

File: block_producer_swap_bot/cli/eth_swap_bot/service.py (hunk 5 above)
Comment (Art17) [reply block_producer_swap_bot/cli/eth_swap_bot/service.py:93]: Done.

File: block_producer_swap_bot/README.md (hunk 4 above)
Comment (Art17) [reply block_producer_swap_bot/README.md]: Done.

File: block_producer_swap_bot/cli/generic/help.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+"""
+Provide help messages for command line interface's commands.
+"""
+
+ETH_PROVIDER_ARGUMENT_HELP_MESSAGE = 'Ethereum provider to pull swap requests from'
+REMNODE_ARGUMENT_HELP_MESSAGE = 'Remnode host and port. For example 127.0.0.1:8888'
+PERMISSION_ARGUMENT_HELP_MESSAGE = 'A block producer account and permission' \
+                                   ' of level to authorize, as in \'account@permission\''
+PRIVATE_KEY_ARGUMENT_HELP_MESSAGE = 'Block producer\'s private key to sign blocks.'
```
Comment (roman-tik) [inline block_producer_swap_bot/cli/generic/help.py]: lets use a separate producer permission to approve swaps

File: block_producer_swap_bot/README.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+```
+
+## Block producers
+
+### Configuration file
+
+Open configuration file.
+
+```bash
+$ nano cli/config.ini
```
Comment (roman-tik) [inline block_producer_swap_bot/README.md]: let's merge swap config with remnode config, so everything is in one place

File: block_producer_swap_bot/README.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+Replace remnode, eth-provider, block-producer-permission, block-producer-private-key with your remnode host and port, a link to Ethereum node with websocket connection, your block producer's account and private key for signing blocks.
+
+
+### Usage
+
+To start approving swaps run the following command
+
+```bash
+$ python3 block_producer_swap_bot.py eth-swap-bot process-swaps
```
Comment (roman-tik) [inline block_producer_swap_bot/README.md]: lets create a common start script, so the operators run a single .sh file to run the node and the swap processors

File: block_producer_swap_bot/cli/generic/help.py (hunk 7 above)
Comment (Art17) [reply block_producer_swap_bot/cli/generic/help.py]: Done.

File: block_producer_swap_bot/README.md (hunk 9 above)
Comment (Art17) [reply block_producer_swap_bot/README.md]: Done.

**Pull request merged**